File: zul/src/org/zkoss/zul/api/Separator.java
/* Separator.java

	Purpose:
		
	Description:
		
	History:
		Tue Oct 22 14:45:31     2008, Created by Flyworld

Copyright (C) 2008 Potix Corporation. All Rights Reserved.

{{IS_RIGHT
	This program is distributed under LGPL Version 3.0 in the hope that
	it will be useful, but WITHOUT ANY WARRANTY.
}}IS_RIGHT
 */
package org.zkoss.zul.api;

import org.zkoss.zk.ui.WrongValueException;

/**
 * A separator.
 * <p>
 * Default {@link #getZclass} as follows: (since 3.5.0)
 * <ol>
 * <li>Case 1: If {@link #getOrient()} is vertical and {@link #isBar()} is
 * false, "z-separator-ver" is assumed</li>
 * <li>Case 2: If {@link #getOrient()} is vertical and {@link #isBar()} is true,
 * "z-separator-ver-bar" is assumed</li>
 * <li>Case 3: If {@link #getOrient()} is horizontal and {@link #isBar()} is
 * false, "z-separator-hor" is assumed</li>
 * <li>Case 4: If {@link #getOrient()} is horizontal and {@link #isBar()} is
 * true, "z-separator-hor-bar" is assumed</li>
 * </ol>
 * 
 * @author tomyeh
 * @since 3.5.2
 */
public interface Separator extends org.zkoss.zul.impl.api.XulElement {

	/**
	 * Returns the orient.
	 * <p>
	 * Default: "horizontal".
	 */
	public String getOrient();

	/**
	 * Sets the orient.
	 * 
	 * @param orient
	 *            either "horizontal" or "vertical".
	 */
	public void setOrient(String orient) throws WrongValueException;

	/**
	 * Returns whether it is a horizontal separator.
	 * 
	 */
	public boolean isHorizontal();

	/**
	 * Returns whether it is a vertical separator.
	 * 
	 */
	public boolean isVertical();

	/**
	 * Returns whether to display a visual bar as the separator.
	 * <p>
	 * Default: false
	 */
	public boolean isBar();

	/**
	 * Sets whether to display a visual bar as the separator.
	 */
	public void setBar(boolean bar);

	/**
	 * Returns the spacing.
	 * <p>
	 * Default: null (depending on CSS).
	 */
	public String getSpacing();

	/**
	 * Sets the spacing.
	 * 
	 * @param spacing
	 *            the spacing (such as "0", "5px", "3pt" or "1em")
	 */
	public void setSpacing(String spacing);
}

File: zul/src/main/resources/web/js/zul/wgt/Separator.ts
/* Separator.ts

	Purpose:

	Description:

	History:
		Wed Nov  5 16:58:56     2008, Created by tomyeh

Copyright (C) 2008 Potix Corporation. All Rights Reserved.

This program is distributed under LGPL Version 2.1 in the hope that
it will be useful, but WITHOUT ANY WARRANTY.
*/

function _shallFixPercent(wgt: zul.wgt.Separator): boolean {
	return !!(zk.gecko && wgt._spacing?.endsWith('%'));
}

/**
 * A separator.
 * @defaultValue {@link getZclass} is "z-separator".
 */
@zk.WrapClass('zul.wgt.Separator')
export class Separator extends zul.Widget {
	/** @internal */
	_orient = 'horizontal';
	/** @internal */
	_bar = false;
	/** @internal */
	_spacing?: string;

	/**
	 * @returns the orient.
	 * @defaultValue `"horizontal"`.
	 */
	getOrient(): string {
		return this._orient;
	}
	/**
	 * Sets the orient.
	 * @param orient - either "horizontal" or "vertical".
	 */
	setOrient(orient: string, opts?: Record<string, boolean>): this {
		const o = this._orient;
		this._orient = orient;
		if (o !== orient || opts?.force) {
			this.updateDomClass_();
		}
		return this;
	}
	/**
	 * @returns whether to display a visual bar as the separator.
	 * @defaultValue `false`
	 */
	isBar(): boolean {
		return this._bar;
	}
	/**
	 * Sets  whether to display a visual bar as the separator.
	 */
	setBar(bar: boolean, opts?: Record<string, boolean>): this {
		const o = this._bar;
		this._bar = bar;
		if (o !== bar || opts?.force) {
			this.updateDomClass_();
		}
		return this;
	}
	/**
	 * @returns the spacing.
	 * @defaultValue `null` (depending on CSS).
	 */
	getSpacing(): string | undefined {
		return this._spacing;
	}
	/**
	 * Sets the spacing.
	 * @param spacing - the spacing (such as "0", "5px", "3pt" or "1em")
	 */
	setSpacing(spacing: string, opts?: Record<string, boolean>): this {
		const o = this._spacing;
		this._spacing = spacing;
		if (o !== spacing || opts?.force) {
			this.updateDomStyle_();
		}
		return this;
	}

	/**
	 * @returns whether it is a vertical separator.
	 */
	isVertical(): boolean {
		return this._orient == 'vertical';
	}

	/** @internal */
	override bind_(desktop?: zk.Desktop, skipper?: zk.Skipper, after?: CallableFunction[]): void {
		super.bind_(desktop, skipper, after);
	}
	override getZclass(): string {
		return 'z-separator';
	}
	/** @internal */
	override domClass_(no?: zk.DomClassOptions): string {
		var sc = super.domClass_(no),
			bar = this.isBar();
		if (!no || !no.zclass) {
			sc += ' ' + this.$s((this.isVertical() ? 'vertical' + (bar ? '-bar' : '') :
				'horizontal' + (bar ? '-bar' : '')));
		}
		return sc;
	}
	/** @internal */
	override domStyle_(no?: zk.DomStyleOptions): string {
		var s = super.domStyle_(no);
		if (!_shallFixPercent(this))
			return s;

		//_spacing contains %
		var space = this._spacing!,
			v = zk.parseInt(space.substring(0, space.length - 1).trim());
		if (v <= 0) return s;
		const percent = `${v >= 2 ? v / 2 : 1}%`;

		return 'margin:' + (this.isVertical() ? '0 ' + percent : percent + ' 0')
			+ ';' + s;
	}
	override getWidth(): string | undefined {
		const wd = super.getWidth();
		return !this.isVertical() || (wd != null && wd.length > 0)
			|| _shallFixPercent(this) ? wd : this._spacing;

	}
	override getHeight(): string | undefined {
		const hgh = super.getHeight();
		return this.isVertical() || (hgh != null && hgh.length > 0)
			|| _shallFixPercent(this) ? hgh : this._spacing;
	}
}